src/install-pnpm/run.ts

```diff
@@ -41,18 +43,40 @@ async function readTarget(opts: {
   readonly standalone: boolean
 }) {
   const { version, packageJsonFile, standalone } = opts
+  const { GITHUB_WORKSPACE } = process.env
 
-  if (version) return `${ standalone ? '@pnpm/exe' : 'pnpm' }@${version}`
+  let packageManager
+
+  if (GITHUB_WORKSPACE) {
+    try {
+      ({ packageManager } = JSON.parse(readFileSync(path.join(GITHUB_WORKSPACE, packageJsonFile), 'utf8')))
+    } catch (error: unknown) {
+      // Swallow error if package.json doesn't exist in root
+      if (!util.types.isNativeError(error) || !('code' in error) || error.code !== 'ENOENT') throw error
+    }
+  }
+
+  if (version) {
+    if (
+      typeof packageManager === 'string' &&
+      packageManager.replace('pnpm@', '') !== version
```

Comment:
this isn't enough, there can be a hash at end. see recent corepack release.

Reply:
Oh sorry for any breakage! Can you provide an example of the hash + delimiter? I can open a new PR.

Reply:
I saw this behavior in Corepack just today, the `packageManager` specifier looks like this:

```json
   "packageManager": "pnpm@9.1.1+sha512.14e915759c11f77eac07faba4d019c193ec8637229e62ec99eefb7cf3c3b75c64447882b7c485142451ee3a6b408059cdfb7b7fa0341b975f12d0f7629c71195"
```

Reply:
@viceice but maybe this is actually desirable behavior, since:

- the `version` input on `pnpm/action-setup` supports a specifier with a Corepack hash (see output of `pnpm install` below)
- technically it could be argued that `9.1.1` is not the same as `9.1.1+sha512.14e915759c11f77eac07faba4d019c193ec8637229e62ec99eefb7cf3c3b75c64447882b7c485142451ee3a6b408059cdfb7b7fa0341b975f12d0f7629c71195`, so stripping the hash is maybe not wanted

```bash
$ pnpm install pnpm@9.1.1+sha512.14e915759c11f77eac07faba4d019c193ec8637229e62ec99eefb7cf3c3b75c64447882b7c485142451ee3a6b408059cdfb7b7fa0341b975f12d0f7629c71195

Packages: +1
+
Progress: resolved 1, reused 0, downloaded 1, added 1, done

dependencies:
+ pnpm 9.1.1

Done in 1.3s
```

Reply:
Thanks, but does it also work with the standalone version?

Reply:
Looks like it, yep:

```bash
$ pnpm install @pnpm/exe@9.1.1+sha512.14e915759c11f77eac07faba4d019c193ec8637229e62ec99eefb7cf3c3b75c64447882b7c485142451ee3a6b408059cdfb7b7fa0341b975f12d0f7629c71195
Packages: +2
++
Downloading @pnpm/macos-arm64@9.1.1: 22.84 MB/22.84 MB, done
Progress: resolved 7, reused 1, downloaded 2, added 2, done
node_modules/.pnpm/@pnpm+exe@9.1.1/node_modules/@pnpm/exe: Running preinstall script, done in 46ms

dependencies:
+ @pnpm/exe 9.1.1

Done in 8.6s
```